tests/results/test_workflow_result.py

```diff
@@ -840,3 +840,36 @@ def test_write_to_pickle_file(self):
         with self.assertRaises(
                 RuntimeError):
             wfr.write_to_pickle_file(None)
+
+    def test_write_to_pickle_file_with_merge(self):
```

Comment:
since this isn't creating threads its not actauly happening in parallel, we will need to create a test that creates multiple threads and causes them to consistently attempt to run in parallel.

Reply:
So I chewed on this for a while...I'm not quite sure how we'd create a multi-threaded test that can reliably test for parallel updates to the pickle file. If the threads end up running sequentially by some fluke of OS scheduling then we don't test what we need to test.

In the end I opted for testing the end result that we're looking for, since this test runs at the `WorkflowResult` level it doesn't care about locking and makes that the caller's problem.

We could add an additional test into `StepRunner` to verify the lock is held. If this was Python 3.8 and later we could watch for audit events generated by flock, or perhaps we just look for the existence of the separate lock file and consider that sufficient? Very keen for ideas :)

Reply:
hmmm. i think we could do it without the threads asctaully.

1) create a mock pickle file (we have utils for that)
2) create a mock StepImpelemnter1 and have it load the pickle file
3) create a mock StepImplementer2 and have it load the pickle file
4) add some results to StepImplementer1
5) add some results to StepImplementer2
6) write StepImpletner1 results to pickle
7) write StepImplementer2 results to pickle

that will test everything but the file locking actually working, which yeah, can't think of a good way to make sure that hapepns, but so long as we lock on the pickle file before the read, merge, write, we should be good enough

Reply:
:+1: sounds good. I'll update the tests here in `WorkflowResult` to cover that and push with the other updates.

Reply:
I've updated the test to reflect this sequence. Let me know what you think.